core/src/main/scala/cats/instances/function.scala

```diff
@@ -71,6 +71,21 @@ sealed private[instances] trait Function0Instances extends Function0Instances0 {
 
       def pure[A](x: A): () => A = () => x
 
+      override def map[A, B](fa: () => A)(fn: A => B): () => B =
+        () => fn(fa())
+
+      override def map2[A, B, C](fa: () => A, fb: () => B)(fn: (A, B) => C): () => C =
+        () => fn(fa(), fb())
+
+      override def product[A, B](fa: () => A, fb: () => B): () => (A, B) =
+        () => (fa(), fb())
+
+      override def ap[A, B](f: () => A => B)(fa: () => A): () => B =
+        () => {
+          val fnAB = f()
```

Comment:
You can write this as either `f.apply.apply(fa())` or `f()(fa())` on Dotty, both of which are I think a little nicer than the extra `val`.